### PR TITLE
fix: transaction not ending correctly

### LINKED ISSUE #11260: Regression in transactionDepth handling?
### Issue description

transactionDepth went negative

### Expected Behavior

transactionDepth is never decremented more often than it was incremented

### Actual Behavior

Negative transaction depth

### Steps to reproduce

Nest a couple of transactions and cause a DB error (e.g. insert a bad value into a column of the wrong type)

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |          |
| Node.js version     | 22.6.0  |
| Typescript version  | 5.4.5  |
| TypeORM version     | 0.3.20  |

### Additional Context

Seen in the wild, in a situation where an unhandled database error occurred deep down in a transaction stack, the transactionDepth appears to be decremented more often than it was incremented. When a new save point is attempted it will cause a syntax error due to the minus sign on the negative number.

Might be related to the changes from https://github.com/typeorm/typeorm/pull/10210

```
      "err": {
        "driverError": {
          "length": 90,
          "position": "19",
          "stack": "error: syntax error at or near \"-\"\n    at /app/node_modules/pg/lib/client.js:526:17\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async PostgresQueryRunner.query (/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:184:25)\n    at async PostgresQueryRunner.startTransaction (/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:130:13)\n    at async EntityPersistExecutor.execute (/app/node_modules/typeorm/persistence/EntityPersistExecutor.js:134:25)",
          "name": "error",
          "routine": "scanner_yyerror",
          "code": "42601",
          "type": "DatabaseError",
          "line": "1180",
          "severity": "ERROR",
          "message": "syntax error at or near \"-\"",
          "file": "scan.l"
        },
        "file": "scan.l",
        "stack": "QueryFailedError: syntax error at or near \"-\"\n    at PostgresQueryRunner.query (/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219:19)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async PostgresQueryRunner.startTransaction (/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:130:13)\n    at async EntityPersistExecutor.execute (/app/node_modules/typeorm/persistence/EntityPersistExecutor.js:134:25)",
        "severity": "ERROR",
        "position": "19",
        "routine": "scanner_yyerror",
        "type": "QueryFailedError",
        "length": 90,
        "code": "42601",
        "message": "syntax error at or near \"-\"",
        "line": "1180",
        "query": "SAVEPOINT typeorm_-7"
      }
```

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [x] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [ ] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, but I don't know how to start. I would need guidance.

### PR DESCRIPTION
**Pull request by alumni**
### Description of change

Fixes #11260.

Reverts #10210 which also introduced the flaky tests - the tests don't actually make sense.

This should also solve the CockroachDB issues we tried to solve in #11190. The main issue with CRDB is that if a query fails, you can't release the savepoint anymore - you can only roll back (see https://www.cockroachlabs.com/docs/stable/release-savepoint#handling-errors).

I also cleaned up two of tests a little bit, they had some unnecessary complexity.

### Pull-Request Checklist

- [X] Code is up-to-date with the `master` branch
- [X] `npm run format` to apply prettier formatting
- [X] `npm run test` passes with this change
- [X] This pull request links relevant issues as `Fixes #11260`
- [X] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [X] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Review by OSA413 — APPROVED** [review]
Looks like you managed to improve the code (the transactions and tests)

**Comment by alumni** [comment]
@OSA413 I added a test for #11260 (fails on master, passes in this branch) and took the PR out of draft.

It can be reviewed now.

File: src/query-builder/RelationIdLoader.ts [hunk 1]
```diff
@@ -636,7 +637,7 @@ export class RelationIdLoader {
                 undefined,
                 primaryColumn.entityMetadata.name +
                     "_" +
-                    relation.inverseRelation!.propertyPath.replace(".", "_") +
+                    originalRelation.propertyPath.replace(".", "_") +
```
Comment (G0maa) [inline src/query-builder/RelationIdLoader.ts:640]: Since this is not part of the revert, what's happening here? :eyes:

**Comment by G0maa** [comment]
For history reasons, this is a related [discussion](https://discord.com/channels/1124626013360488529/1336686763464200274) on TypeORM Discord.

File: src/query-builder/RelationIdLoader.ts (hunk 1 above)
Comment (alumni) [reply src/query-builder/RelationIdLoader.ts:640]: A test was failing because the value was undefined.

**Review by G0maa — APPROVED** [review]

**Review by sgarner — APPROVED** [review]
The code changes look fine. :+1:

Are we happy that this will reintroduce the bug described in #10210? It seems to me that submitting queries in parallel to a single transaction was always an incorrect practice which shouldn't be supported. But if anyone is relying on this behavior, reverting might represent a breaking regression.

**Comment by alumni** [comment]
> Are we happy that this will reintroduce the bug described in #10210? It seems to me that submitting queries in parallel to a single transaction was always an incorrect practice which shouldn't be supported. But if anyone is relying on this behavior, reverting might represent a breaking regression.

@sgarner #10210 is not a bug, it is introducing a bug. Reverting that PR that introduced that bug will fix the regression introduced by it.

**Review by naorpeled — APPROVED** [review]

**Comment by sgarner** [comment]
> @sgarner #10210 is not a bug, it is introducing a bug. Reverting the related PR will fix the regression introduced by it.

#10210 was fixing https://github.com/typeorm/typeorm/issues/10209

If you're arguing that #10209 wasn't a bug that needed fixing, then I'm in agreement with you. But we at least ought to note this in the changelog.

**Comment by linbrandon4** [comment]
This is a big one, but the direction makes sense. Reverting #10210 to fix the transaction regression + flaky tests feels justified, especially since the “fix” was causing more problems than it solved. The CockroachDB angle also checks out — once a query fails, you can’t just release the savepoint, you have to roll back, so the old behavior was basically fighting the DB.

A few things I’d focus on if I were reviewing closely:
- Behavior change / possible breaking impact: the discussion about parallel queries inside a single transaction is important. If we’re deciding “promise.all inside one transaction isn’t supported,” that should be clearly stated somewhere (changelog/docs), because some users might have been relying on it even if it was never a good idea.
- Nested transaction correctness: make sure transactionDepth and savepoint naming/rollback behavior is consistent across drivers (Postgres vs Cockroach vs SQLite etc.), especially after a failure mid-transaction.
- Tests: since this PR touches 25 files and drops a ton of lines, the tests are basically the safety net. I like that there’s a test that fails on master and passes here (#11260) — I’d just want to confirm the “improved” tests actually reflect real driver behavior and aren’t masking edge cases.
